### DISCUSSION TITLE
Dcm4che-net - Preferred Transfer Syntax selection algorithm applies on each Presentation Context instead of on each SOP Class

### DISCUSSION THREAD

**Discussion by ChrisP974** [dcm4che/dcm4che#1137] [Q&A]
I have this issue regarding the associations negotiations on C-STORE and the way it works in Dcm4chee-net.
 
More precisely about the Presentation Context negotiations, and the choice of Transfer Syntax, when we have some "preferred transfer syntax" that are configured.
 
I want to be able to prioritize a TranferSyntax, without having to no longer support others.
 
Let say I have this configuration, for dcm4che as a C-STORE-SCP (the receiving one - regardless of the SOP Class) :
"transfertSyntaxSupported": [
   "1.2.840.10008.1.2.1",
   "1.2.840.10008.1.2.2",
   "1.2.840.10008.1.2",
   "1.2.840.10008.1.2.4.90"
],
"preferredTransferSyntax": [
   "1.2.840.10008.1.2.2"
]
 
Let's suppose we have a SCU (named SCU_1) that proposes one transfer syntax per presentation context :
PresentationContext1:
   SOPClass: CT Image Storage
   TransferSyntax: 1.2.840.10008.1.2.4.90
PresentationContext2:
   SOPClass: CT Image Storage
   TransferSyntax: 1.2.840.10008.1.2.2
 
And then another SCU (named SCU_2) that proposes many transfer syntax per presentation context:
PresentationContext1:
   SOPClass: CT Image Storage
   TransferSyntax: 1.2.840.10008.1.2.4.90, 1.2.840.10008.1.2.2
PresentationContext2:
   SOPClass: CT Image Storage
   TransferSyntax: 1.2.840.10008.1.2  
 
During negotiations, the SCU exposes the PresentationContext, and the SCP tells if each PC is Accepted/Refused and for each accepted, one selected transfer syntax. 
 
**With my configuration, here is what I am expecting :**
For SCU_1:
   PresentationContext1:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.4.90
      Result: Refused (because another PC contains the preferred TS) 
   PresentationContext2:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.2
      Result: Accepted with TransferSyntax 1.2.840.10008.1.2.2 (preferred one)
For SCU_2:
   PresentationContext1:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.4.90, 1.2.840.10008.1.2.2
      Result: Accepted with TransferSyntax 1.2.840.10008.1.2.2 (the preferred one)
   PresentationContext2:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2 
      Result: Refused (because another PC contains the preferred TS)
 
**However, here is the issue I observe :**
For SCU_1:
   PresentationContext1:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.4.90
      Result: Accepted (because it is supported)
   PresentationContext2:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.2
      Result: Accepted with TransferSyntax 1.2.840.10008.1.2.2 (because supported)
For SCU_2:
   PresentationContext1:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2.4.90, 1.2.840.10008.1.2.2
      Result: Accepted with TransferSyntax 1.2.840.10008.1.2.2 (because supported and preferred to the other TS)
   PresentationContext2:
      SOPClass: CT Image Storage
      TransferSyntax: 1.2.840.10008.1.2  (because another PC contains the preferred TS)
      Result: Accepted (because supported)
 
**In summary**, I observe that the "preferred transfer syntax" logic only applies on each PresentationContext. But as many PresentationContext are accepted, the SCU will use one of them, and not necessarily the one with the preferred transfer syntax. What I am expecting is that, for a given SOP Class, the SCP accepts only the PresentationContext which contains the preferredTransferSyntax, and refuse the other one. This way the SCU will be forced to use the SCP's preferred Transfer Syntax. 

Here is the code where I see this issue, in AssociationHandler.makeAAssociateAC :
![image](https://user-images.githubusercontent.com/66563302/157212069-c8ad82cd-0354-45bb-b330-14a9d9b69997.png)
![image](https://user-images.githubusercontent.com/66563302/157212252-d10069dd-3442-4cf7-a49e-13dc5574d884.png)
![image](https://user-images.githubusercontent.com/66563302/157212274-4cde8a65-c28d-4309-bc8e-70f114ad70ec.png)

**Comment by gunterze** [comment]
This behavior is intended. "Prefer" does not mean to reject PCs with only offers not-preferred TCs.

**Comment by ChrisP974** [comment]
Thanks Gunter for your answer

So, today in dcm4che there is no way to force a Transfer Syntax in the case when the SCU proposes many PCs for a single SOPClass ?

**Comment by gunterze** [comment]
You can't configure conditional rejection, dependent on if a "preferred" TC for the same AS was offered  in another PC .

**Comment by aboufahj** [comment]
Hi. Maybe we can think about a new "contextual preferred" TransferSyntaxes property, which have the boundary of the preference linked to the whole association negotiation, not only to the PC boundaries.
